### PR TITLE
`sol!` macro examples

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

need `sol!` macro examples covering its various abilities.

## Solution

- [x] Types and UDVTs
- [x] Struct and Enums
- [x] Contract interaction from solidity and json abi 
- [x] Events, Functions and Error
## PR Checklist

- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Comment by yash-atreya** [comment]
@DaniPopes @mattsse Any specific `sol!` macro abilities you'd like to be demonstrated in the examples apart from the ones listed above?

**Comment by yash-atreya** [comment]
Will add an example to depict rust attributes inside the macro

**Comment by zerosnacks** [comment]
Any blockers?

Recently added https://alloy.rs/highlights/the-sol!-procedural-macro.html, could perhaps be useful for practical examples

**Comment by yash-atreya** [comment]
> Any blockers?
> 
> Recently added https://alloy.rs/highlights/the-sol!-procedural-macro.html, could perhaps be useful for practical examples that we can then embed into the page

Sorry for the delay.
Just cleaning up. Do you have ideas/powerful usecases that you depicted in the docs? 

I can implement those.

File: examples/sol-macro/examples/udvt_aliases.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+//! Examples showing defining user defined value types (UDVTs) and type aliases using the sol macro
```
Comment (zerosnacks) [inline examples/sol-macro/examples/udvt_aliases.rs:1]: Perhaps preferable to rename file to `user_defined_types.rs`

**Comment by zerosnacks** [comment]
> > Any blockers?
> > Recently added https://alloy.rs/highlights/the-sol!-procedural-macro.html, could perhaps be useful for practical examples that we can then embed into the page
> 
> @zerosnacks Sorry for the delay. Just cleaning up. Do you have ideas/powerful usecases that you depicted in the docs?
> 
> I can implement those.

I think what you have now is sufficient, though having the events (as shown in https://docs.rs/alloy-sol-macro/latest/alloy_sol_macro/macro.sol.html) would be nice.

One area we haven't explored a lot is `sol!` <> `alloy-dyn-abi`, that is dynamic Solidity encoding

Perhaps @DaniPopes has specific ideas on what he wants to be highlighted (best practices, limitations, features)

**Comment by yash-atreya** [comment]
> > > Any blockers?
> > > Recently added https://alloy.rs/highlights/the-sol!-procedural-macro.html, could perhaps be useful for practical examples that we can then embed into the page
> > 
> > 
> > @zerosnacks Sorry for the delay. Just cleaning up. Do you have ideas/powerful usecases that you depicted in the docs?
> > I can implement those.
> 
> I think what you have now is sufficient, though having the events (as shown in https://docs.rs/alloy-sol-macro/latest/alloy_sol_macro/macro.sol.html) would be nice.
> 
> One area we haven't explored a lot is `sol!` <> `alloy-dyn-abi`, that is dynamic Solidity encoding
> 
> Perhaps @DaniPopes has specific ideas on what he wants to be highlighted (best practices, limitations, features)

Cool 

Adding events and errors in the next push

**Review by mattsse — APPROVED** [review]

**Review by zerosnacks — APPROVED** [review]
Thanks! lgtm

**Pull request merged**